src/awkward/_v2/cpp-headers/GrowableBuffer.h

```diff
@@ -0,0 +1,263 @@
+    /// allocations of new panels.
+    GrowableBuffer(size_t initial,
+                   std::unique_ptr<PRIMITIVE> ptr,
+                   size_t length,
+                   size_t reserved)
+        : initial_(initial) {
+      ptr_.push_back(std::move(ptr));
+      length_.push_back(length);
+      reserved_.push_back(reserved);
+    }
+
+    /// @brief Creates a GrowableBuffer by allocating a new buffer, taking an
+    /// initial #reserved from #initial.
+    GrowableBuffer(size_t initial)
+        : GrowableBuffer(initial,
+                         std::unique_ptr<PRIMITIVE>(new PRIMITIVE[initial]),
+                         0,
+                         initial) { }
+
+    /// @brief Currently used number of elements.
+    ///
+    /// Although the #length increments every time #append is called,
+    /// it is always less than or equal to #reserved because of
+    /// allocations of new panels.
+    size_t
+    length() const {
+      return std::accumulate(length_.begin(), length_.end(), (size_t)0);
+    }
+
+    /// @brief Currently allocated number of elements.
+    ///
+    /// Although the #length increments every time #append is called,
+    /// it is always less than or equal to #reserved because of
+    /// allocations of new panels.
+    size_t
+    reserved() const {
+      return std::accumulate(reserved_.begin(), reserved_.end(), (size_t)0);
+    }
+
+    /// @brief Discards accumulated data, the #reserved returns to
+    /// initial, and a new #ptr is allocated.
+    void
+    clear() {
+      length_.clear();
+      length_.push_back(0);
+      reserved_.clear();
+      reserved_.push_back(initial_);
+      ptr_.clear();
+      ptr_.push_back(std::unique_ptr<PRIMITIVE>(new PRIMITIVE[initial_]));
+    }
+
+    /// @brief Inserts one `datum` into the panel.
+    void
+    fill_panel(PRIMITIVE datum) {
+      if (length_[ptr_.size()-1] < reserved_[ptr_.size()-1]) {
+        ptr_[ptr_.size()-1].get()[length_[ptr_.size()-1]] = datum;
+        length_[ptr_.size()-1]++;
+      }
+    }
+
+    /// @brief Creates a new panel with slots equal to #reserved.
+    void
+    add_panel(size_t reserved) {
+      ptr_.push_back(std::unique_ptr<PRIMITIVE>(new PRIMITIVE[reserved]));
+      length_.push_back(0);
+      reserved_.push_back(reserved);
+    }
+
+    /// @brief Inserts one `datum` into the panel, possibly triggering
+    /// allocation of a new panel.
+    ///
+    /// This increases the #length by 1; if the new #length is larger than
+    /// #reserved, a new panel will be allocated.
+    void
+    append(PRIMITIVE datum) {
+      if (length_[ptr_.size()-1] == reserved_[ptr_.size()-1]) {
+        add_panel(reserved_[ptr_.size()-1]);
+      }
+      fill_panel(datum);
+    }
+
+    /// @brief Inserts an entire array into the panel(s), possibly triggering
+    /// allocation of a new panel.
+    void
+    append(PRIMITIVE* ptr, size_t size) {
+      for (size_t i = 0; i < size; i++) {
+        append(ptr[i]);
```

Comment:
`append` would trigger if on each element. Would it be better to check the size beforehand and allocate a new panel (possibly a `size` size if it is bigger that `reserved`?